docs/design/simplified_federation_deployment.md

```diff
@@ -0,0 +1,387 @@
+      - [Cluster Deregistration Experience](#cluster-deregistration-experience)
+    - [Points for Discussion:](#points-for-discussion)
+  - [Design](#design)
+      - [Federation Control Plane Bootstrap](#federation-control-plane-bootstrap)
+        - [Additional enhancements](#additional-enhancements)
+        - [Caveats](#caveats)
+      - [Cluster Registration/Deregistration](#cluster-registrationderegistration)
+        - [Registration](#registration)
+        - [Deregistration](#deregistration)
+  - [Test Plan](#test-plan)
+      - [Federation Control Plane Bootstrap](#federation-control-plane-bootstrap-1)
+      - [Cluster Registration/Deregistration](#cluster-registrationderegistration-1)
+  - [Timeline](#timeline)
+  - [Maturity Level](#maturity-level)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+## Background
+
+Kubernetes Cluster Federation has its own control plane which is similar
+to the control plane of a Kubernetes cluster. Like Kubernetes, a
+federation control plane is composed of an API server and a controller
+manager that bundles all the core controllers. While that looks simple
+on paper, deploying a federation control plane comes with its own set of
+challenges. Deploying a federation control plane not just involves
+deploying a federation API server and a federation controller manager,
+but also other components that support the control plane. This includes:
+
+1. A persistent store to store the state of the federation - `etcd`.
+2. A persistent device to store `etcd` data.
+3. A secret store to store the federation API server credentials.
+
+And these components must be setup in such a way that they can
+appropriately coordinate with each other. While this is already starting
+to look complicated, this is only one part of the problem.
+
+Kubernetes Cluster Federation operates only on the clusters that it is
+responsible for. So once a federation control plane is deployed, for it
+to be useful, users should be able to register and deregister Kubernetes
+clusters with/from their federation. In order to register a cluster,
+users should supply the cluster's credentials and its API server
+endpoint.
+
+**Side note**: It is a mouthful to always prefix "federation" to each of
+its control plane components. But it is necessary to disambiguate these
+components from the almost identical components in a Kubernetes cluster.
+
+## Goals
+
+Today, we ship bash scripts to facilitate the federation control plane
+deployment process and the registration/deregistration process is mostly
+manual. Both these processes involve multiple coordinated steps and are
+error prone. For example, to register a Kubernetes cluster with
+federation users have to write a yaml file describing a cluster
+resource. They have to specify the cluster's endpoint in the right
+format. That field is too restrictive and hard to get it right. They
+also have to correctly specify where the credentials for the cluster is
+stored.
```

Comment:
I say so, because I don't see a direct/quick method of building and deploying federation images, when changing/extending src (atleast here). Am I missing something? (Please correct me, if I am assuming that the method seems to use the release hyperkube images). Also, as listed in this doc, the kubeadm init federation seems to initialize the base cluster also, m not sure if there can be hooks for a shorter process (for example cluster already exists and we need initialize federation components only on top); this might however purely be choice of the implementation though.